src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java

```diff
@@ -486,42 +488,61 @@ private void handleCurrentStateBroken(Optional<State> desiredState, State previo
     }
 
     @SuppressWarnings("PMD.AvoidCatchingThrowable")
-    private void handleCurrentStateNew(Optional<State> desiredState) throws InterruptedException {
+    private void handleCurrentStateNewAsync(Optional<State> desiredState,
+                                       AtomicReference<Predicate<Object>> asyncFinishAction) {
         // if no desired state is set, don't do anything.
         if (!desiredState.isPresent()) {
             return;
         }
 
+        // if there is already a install() task running, do nothing.
+        Future<?> currentTask = backingTask.get();
+        if (currentTask != null && !currentTask.isDone()) {
+            return;
+        }
+
         long currentStateGeneration = stateGeneration.incrementAndGet();
+
+        Integer timeout = getTimeoutConfigValue(
+                LIFECYCLE_INSTALL_NAMESPACE_TOPIC, DEFAULT_INSTALL_STAGE_TIMEOUT_IN_SEC);
+        Future<?> schedule =
+                greengrassService.getContext().get(ScheduledExecutorService.class).schedule(() -> {
+                    if (getState().equals(State.NEW) && currentStateGeneration == getStateGeneration().get()) {
+                        greengrassService.serviceErrored(ComponentStatusCode.INSTALL_TIMEOUT, "Timeout in install");
+                    }
+                }, timeout, TimeUnit.SECONDS);
+
         replaceBackingTask(() -> {
             if (!State.NEW.equals(getState()) || getStateGeneration().get() != currentStateGeneration) {
                 // Bail out if we're not in the expected state
                 return;
             }
             try {
                 greengrassService.install();
+                if (!State.ERRORED.equals(lastReportedState.get()) && !schedule.isDone()) {
```

Comment:
you don't need to check `!schedule.isDone()` because state will be ERRORED if schedule timed out

Reply:
The unit tests related to the installation timeout cannot pass if we don't check it.